PR: https://github.com/odoo/odoo/pull/99023

From: 6927dcda9f99db2a4d2f06b32f8d377c17fe2794
From: eres-odoo
Date: 2022-09-02 18:10:59

Structural Changes: 7
Total Changes: 9

[IMP] loyalty*: misc loyalty improvements

This commit improves some messages and usability for loyalty and its
submodules:
 - Coupons should not be restricted to the partner assigned to them
   anymore
 - Merged "I have a promo code" and "Use a gift card" into a single
   toggle
 - Boolean fields have been added to each app to disable them where we
   don't want them.
 - Reduced the size of coupon codes
   This means that we increase the risk of collision which should be
   handled by the model as well now.
   Creating a coupon with an existing code will retry the creation (up
   to 10 times) until the coupon is indeed created.
 - Make it clearer in the coupon generation wizard that an email will be
   sent to the customer.
 - Reintroduce coupon/gift card details on sale order lines in eCommerce
   which was wrongfully removed.
 - Remove the "on product with taxes: " message on order lines when
   the amount is not split between taxes or if the tax message would be
   empty.
 - Make it possible for some error messages to be displayed as warning
   instead of errors
 - Split program types in two menuitems, one for promotions discount
   loyalty etc and one for gift cards and ewallet, the latter has
   simplified views.
 - Views overall have been reviewed to make it easier to understand how
   to programs work. New program types have also been added.

TaskId-2951413

Part-of: odoo/odoo#99023
Co-authored-by: William Braeckman <wbr@odoo.com>

================================= pseudo patch: =================================

--- a/addons/pos_loyalty/models/loyalty_program.py
+++ b/addons/pos_loyalty/models/loyalty_program.py
@@ -9,6 +9,7 @@ class LoyaltyProgram(models.Model):
 
     pos_config_ids = fields.Many2many('pos.config', string="Point of Sales", readonly=True)
     pos_order_count = fields.Integer("PoS Order Count", compute='_compute_pos_order_count')
+    pos_ok = fields.Boolean("Point of Sale", default=True)
 
     def _compute_pos_order_count(self):
         read_group_res = self.env['pos.order.line']._read_group(

--- a/addons/pos_loyalty/models/pos_config.py
+++ b/addons/pos_loyalty/models/pos_config.py
@@ -13,16 +13,16 @@ class PosConfig(models.Model):
     use_coupon_programs = fields.Boolean('Coupons & Promotions',
         help="Use coupon and promotion programs in this PoS configuration.")
     coupon_program_ids = fields.Many2many(
-        'loyalty.program', string="Coupon Programs", domain=[('program_type', '=', 'coupons')],
+        'loyalty.program', string="Coupon Programs", domain=[('pos_ok', '=', True), ('program_type', '=', 'coupons')],
         relation='pos_config_coupon_program_rel')
     promo_program_ids = fields.Many2many(
-        'loyalty.program', string="Promotion Programs", domain=[('program_type', '=', 'promotion')],
+        'loyalty.program', string="Promotion Programs", domain=[('pos_ok', '=', True), ('program_type', 'in', ('promotion', 'buy_x_get_y', 'promo_code', 'next_order_coupons'))],
         relation='pos_config_promo_program_rel')
 
-    loyalty_program_id = fields.Many2one('loyalty.program', "PoS Loyalty Programs", domain=[('program_type', '=', 'loyalty')], default=_default_loyalty_program)
+    loyalty_program_id = fields.Many2one('loyalty.program', "PoS Loyalty Programs", domain=[('pos_ok', '=', True), ('program_type', '=', 'loyalty')], default=_default_loyalty_program)
 
     use_gift_card = fields.Boolean('Gift Cards')
-    gift_card_program_id = fields.Many2one('loyalty.program', "PoS Gift Card Program", domain=[('program_type', '=', 'gift_card')])
+    gift_card_program_id = fields.Many2one('loyalty.program', "PoS Gift Card Program", domain=[('pos_ok', '=', True), ('program_type', '=', 'gift_card')])
     gift_card_settings = fields.Selection(
         [
             ("create_set", "Generate a new barcode and set a price"),
